## Import libraries

In [1]:
import pandas as pd
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import CountVectorizer
import re
from nltk.stem import WordNetLemmatizer
from nltk.tokenize import RegexpTokenizer


In [2]:
pd.set_option('display.max_columns', 200)
pd.set_option('display.max_rows', 900)
pd.set_option('display.max_colwidth', 200)

In [3]:
sherlock_df = pd.read_csv('../datasets/sherlock.csv')

In [4]:
elementary_df = pd.read_csv('../datasets/elementary.csv')     

In [5]:
sherlock_df.head()

approved_at_utc subreddit  \
0              NaN  Sherlock   
1              NaN  Sherlock   
2              NaN  Sherlock   
3              NaN  Sherlock   
4              NaN  Sherlock   

                                                                                                                                                                                                  selftext  \
0                                                                                                                                                                                                      NaN   
1  Watching aSiB last night, wondered what the writing in the picture above Sherlock's bed was about (I mistakenly thought it was Chinese (it's Japanese) because it was written vertical, but apparent...   
2  How was Charles Augustus Magnussen able to read people's details and pressure points? At first, it seems it was because of his glasses, but then we see him reading Sherlock without them. Did he ge...   
3                                                                                                                                                                                                      NaN   
4                                                                                                                                                                                                      NaN   

  author_fullname  saved  mod_reason_title  gilded  clicked  \
0     t2_55taapag  False               NaN       0    False   
1     t2_4k9thwt8  False               NaN       0    False   
2     t2_2voyl5a8  False               NaN       0    False   
3     t2_55ula4y0  False               NaN       0    False   
4     t2_5qim4oqt  False               NaN       0    False   

                                                            title  \
0                                    Guys is this Martin freeman?   
1  Mark Gatiss Twitter Response to a Question on One of the Props   
2                            Spoilers for His Last Vow (S03, E03)   
3                                            Sounds like Sherlock   
4               I need a word to describe Mrs Hudson, ie a Legend   

  link_flair_richtext subreddit_name_prefixed  hidden  pwls  \
0                  []              r/Sherlock   False     6   
1                  []              r/Sherlock   False     6   
2                  []              r/Sherlock   False     6   
3                  []              r/Sherlock   False     6   
4                  []              r/Sherlock   False     6   

  link_flair_css_class  downs  thumbnail_height  hide_score       name  \
0                Image      0             140.0       False  t3_glkhwz   
1           Discussion      0             108.0       False  t3_glj1dv   
2           Discussion      0               NaN       False  t3_glifs2   
3                Image      0             140.0       False  t3_gle5qd   
4           Discussion      0               NaN       False  t3_gldsbh   

   quarantine link_flair_text_color  upvote_ratio  \
0       False                  dark          0.72   
1       False                  dark          0.88   
2       False                  dark          0.50   
3       False                  dark          0.98   
4       False                  dark          0.50   

   author_flair_background_color subreddit_type  ups  total_awards_received  \
0                            NaN         public    3                      0   
1                            NaN         public    6                      0   
2                            NaN         public    0                      0   
3                            NaN         public  654                      0   
4                            NaN         public    0                      0   

  media_embed  thumbnail_width author_flair_template_id  is_original_content  \
0          {}            140.0                      NaN                False   
1

In [6]:
elementary_df.head()

approved_at_utc   subreddit  \
0              NaN  elementary   
1              NaN  elementary   
2              NaN  elementary   
3              NaN  elementary   
4              NaN  elementary   

                                                                                                                                                                                                  selftext  \
0                                                                                                                                                                               CLYDE. I love him so much.   
1  This may seem oddly specific and really arbitrary, but I'm trying to figure out all the scenes when Sherlock/JLM is exercising/working out.\n\nThink when he is doing crunches, or push ups, or yoga...   
2                                                                                                                                                                                                      NaN   
3                                                                                                                                                     I know its obvious, and on purpose, but I love them.   
4  We're on our rewatch of the show during lockdown and it's become a game between us to find the [fish vase](https://images.app.goo.gl/EkHKrMnAj5vHP5pr8) which has been repeating itself in episode b...   

  author_fullname  saved  mod_reason_title  gilded  clicked  \
0     t2_5uvk8v4i  False               NaN       0    False   
1      t2_ldj8fxj  False               NaN       0    False   
2     t2_33q0xera  False               NaN       0    False   
3        t2_ytudt  False               NaN       0    False   
4        t2_v4wd8  False               NaN       0    False   

                                                                                                                    title  \
0                                                                                             Best detective on the show?   
1                                                                                                 Sherlock/JLM exercising   
2                                                 Sherlock Stops A Taxi To Use It As A Ladder LMAO [S02E02 - Solve For X]   
3  Some of the most amusing, and touching moments are when Sherlock gets involved in the other characters personal lives.   
4                                                        Has anyone got to the bottom of the reappearing green fish vase?   

  link_flair_richtext subreddit_name_prefixed  hidden  pwls  \
0                  []            r/elementary   False     7   
1                  []            r/elementary   False     7   
2                  []            r/elementary   False     7   
3                  []            r/elementary   False     7   
4                  []            r/elementary   False     7   

   link_flair_css_class  downs  thumbnail_height  hide_score       name  \
0                   NaN      0               NaN       False  t3_gjfniq   
1                   NaN      0               NaN       False  t3_giylxm   
2                   NaN      0              78.0       False  t3_giox5j   
3                   NaN      0               NaN       False  t3_gikndd   
4                   NaN      0               NaN       False  t3_ghkyw2   

   quarantine link_flair_text_color  upvote_ratio  \
0       False                  dark          0.93   
1       False                  dark          0.96   
2       False                  dark          1.00   
3       False                  dark          0.98   
4       False                  dark          0.97   

   author_flair_background_color subreddit_type  ups  total_awards_received  \
0                            NaN         public   30                      0   
1                            NaN         public   18                      0   
2                            NaN         publ

In [7]:
sherlock_df.shape

(1230, 109)

In [8]:
elementary_df.shape

(1242, 109)

In [9]:
sherlock_df.describe()

approved_at_utc  mod_reason_title       gilded    pwls   downs  \
count              0.0               0.0  1230.000000  1230.0  1230.0   
mean               NaN               NaN     0.002439     6.0     0.0   
std                NaN               NaN     0.049346     0.0     0.0   
min                NaN               NaN     0.000000     6.0     0.0   
25%                NaN               NaN     0.000000     6.0     0.0   
50%                NaN               NaN     0.000000     6.0     0.0   
75%                NaN               NaN     0.000000     6.0     0.0   
max                NaN               NaN     1.000000     6.0     0.0   

       thumbnail_height  upvote_ratio  author_flair_background_color  \
count        752.000000   1230.000000                            0.0   
mean         124.059840      0.912098                            NaN   
std           24.330511      0.098256                            NaN   
min           36.000000      0.220000                            NaN   
25%          105.000000      0.880000                            NaN   
50%          140.000000      0.940000                            NaN   
75%          140.000000      0.980000                            NaN   
max          140.000000      1.000000                            NaN   

               ups  total_awards_received  thumbnail_width  category  \
count  1230.000000            1230.000000            752.0       0.0   
mean    192.740650               0.013008            140.0       NaN   
std     330.154019               0.113355              0.0       NaN   
min       0.000000               0.000000            140.0       NaN   
25%      11.000000               0.000000            140.0       NaN   
50%      29.000000               0.000000            140.0       NaN   
75%     179.250000               0.000000            140.0       NaN   
max    2546.000000               1.000000            140.0       NaN   

             score  approved_by  content_categories  mod_note       created  \
count  1230.000000          0.0                 0.0       0.0  1.230000e+03   
mean    192.740650          NaN                 NaN       NaN  1.583882e+09   
std     330.154019          NaN                 NaN       NaN  5.130056e+06   
min       0.000000          NaN                 NaN       NaN  1.571866e+09   
25%      11.000000          NaN                 NaN       NaN  1.579983e+09   
50%      29.000000          NaN                 NaN       NaN  1.585812e+09   
75%     179.250000          NaN                 NaN       NaN  1.588164e+09   
max    2546.000000          NaN                 NaN       NaN  1.589769e+09   

          wls  removed_by_category  banned_by  likes  suggested_sort  \
count  1230.0                  0.0        0.0    0.0             0.0   
mean      6.0                  NaN        NaN    NaN             NaN   
std       0.0                  NaN        NaN    NaN             NaN   
min       6.0                  NaN        NaN    NaN             NaN   
25%       6.0                  NaN        NaN    NaN             NaN   
50%       6.0                  NaN        NaN    NaN             NaN   
75%       6.0                  NaN        NaN    NaN             NaN   
max       6.0                  NaN        NaN    NaN             NaN   

       banned_at_utc  view_count  author_flair_text  removed_by  num_reports  \
count            0.0         0.0                0.0         0.0          0.0   
mean             NaN         NaN                NaN         NaN          NaN   
std              NaN         NaN                NaN         NaN          NaN   
min              NaN         NaN                NaN         NaN          NaN   
25%              NaN         NaN                NaN         NaN          NaN   
50%              NaN         NaN                NaN         NaN          NaN   
75%              NaN         NaN                NaN         NaN          NaN   
max              NaN         NaN                Na

In [10]:
elementary_df.shape

(1242, 109)

Only the the columns 'selftext' and 'title' contain sufficient amount of text for classification purposes. The other columns do not have information relevant to the problem statement and will be dropped.

In [11]:
sherlock_df= sherlock_df[['selftext', 'title']]

In [12]:
elementary_df= elementary_df[['selftext', 'title']]

In [13]:
sherlock_df.head()

selftext  \
0                                                                                                                                                                                                      NaN   
1  Watching aSiB last night, wondered what the writing in the picture above Sherlock's bed was about (I mistakenly thought it was Chinese (it's Japanese) because it was written vertical, but apparent...   
2  How was Charles Augustus Magnussen able to read people's details and pressure points? At first, it seems it was because of his glasses, but then we see him reading Sherlock without them. Did he ge...   
3                                                                                                                                                                                                      NaN   
4                                                                                                                                                                                                      NaN   

                                                            title  
0                                    Guys is this Martin freeman?  
1  Mark Gatiss Twitter Response to a Question on One of the Props  
2                            Spoilers for His Last Vow (S03, E03)  
3                                            Sounds like Sherlock  
4               I need a word to describe Mrs Hudson, ie a Legend

In [14]:
elementary_df.head()

selftext  \
0                                                                                                                                                                               CLYDE. I love him so much.   
1  This may seem oddly specific and really arbitrary, but I'm trying to figure out all the scenes when Sherlock/JLM is exercising/working out.\n\nThink when he is doing crunches, or push ups, or yoga...   
2                                                                                                                                                                                                      NaN   
3                                                                                                                                                     I know its obvious, and on purpose, but I love them.   
4  We're on our rewatch of the show during lockdown and it's become a game between us to find the [fish vase](https://images.app.goo.gl/EkHKrMnAj5vHP5pr8) which has been repeating itself in episode b...   

                                                                                                                    title  
0                                                                                             Best detective on the show?  
1                                                                                                 Sherlock/JLM exercising  
2                                                 Sherlock Stops A Taxi To Use It As A Ladder LMAO [S02E02 - Solve For X]  
3  Some of the most amusing, and touching moments are when Sherlock gets involved in the other characters personal lives.  
4                                                        Has anyone got to the bottom of the reappearing green fish vase?

## Data Cleaning


### Checking for null values

In [15]:
sherlock_df.isnull().sum()

selftext    793
title         0
dtype: int64

There are 793 null values in the column 'selftext'. Since there are no null values in 'title', the null values will be converted to empty space so that information from 'title' will not be lost

In [16]:
sherlock_df.fillna('',inplace = True)

In [17]:
sherlock_df.isnull().sum()

selftext    0
title       0
dtype: int64

In [18]:
elementary_df.isnull().sum()

selftext    317
title         0
dtype: int64

There are 317 null values in the column 'selftext'. Since there are no null values in 'title', the null values will be converted to empty space so that information from 'title' will not be lost

In [19]:
elementary_df.fillna('',inplace = True)

In [20]:
elementary_df.isnull().sum()

selftext    0
title       0
dtype: int64

### Checking for duplicated posts

In [21]:
sherlock_df['is_duplicate'] =sherlock_df['title'].duplicated()


In [22]:
sherlock_df.is_duplicate.value_counts()

False    971
True     259
Name: is_duplicate, dtype: int64

There are 970 unique Sherlock posts.

In [23]:
elementary_df['is_duplicate'] =elementary_df['title'].duplicated()

In [24]:
elementary_df.is_duplicate.value_counts()

False    989
True     253
Name: is_duplicate, dtype: int64

There are 989 unique Elementary posts.

In [25]:
# dropping duplicate rows

sherlock_df =sherlock_df[sherlock_df.is_duplicate != True]

In [26]:
elementary_df =elementary_df[elementary_df.is_duplicate != True]

In [27]:
sherlock_df.is_duplicate.value_counts()

False    971
Name: is_duplicate, dtype: int64

elementary_df.is_duplicate.value_counts()

In [28]:
# dropping 'is_duplicate' column

In [29]:
sherlock_df.drop(['is_duplicate'], axis =1, inplace = True)


In [30]:
elementary_df.drop(['is_duplicate'], axis =1, inplace = True)

### Combining both columns to have all text data in 1 column 

In [31]:
sherlock_df['combo'] = sherlock_df['title'] + sherlock_df['selftext']

In [32]:
sherlock_df= sherlock_df[['combo']] 

In [33]:
elementary_df['combo'] = elementary_df['title'] + elementary_df['selftext']

In [34]:
elementary_df= elementary_df[['combo']] 

### Removing urls as they are noise

In [35]:
sherlock_df['combo'] = sherlock_df['combo'].replace(to_replace=r'^https?:\/\/.*[\r\n]*',value='',regex=True)

In [36]:
sherlock_df['combo'] = sherlock_df['combo'].replace(to_replace=r'http\S+', value='', regex=True).replace(to_replace=r'www\S+',value= '', regex=True)

In [37]:
elementary_df['combo'] = elementary_df['combo'].replace(to_replace=r'^https?:\/\/.*[\r\n]*',value='',regex=True)

In [38]:
elementary_df['combo']= elementary_df['combo'].replace(to_replace=r'http\S+', value='', regex=True).replace(to_replace=r'www\S+',value= '', regex=True)

### Lemmatizing the words to reduce them to only root words

In [39]:
# Instantiate lemmatizer. (Same as above.)
lemmatizer = WordNetLemmatizer()

In [40]:
# Instantiate Tokenizer
tokenizer = RegexpTokenizer(r'\w+')

In [41]:
lemma_s = []

for words in sherlock_df['combo']:
    word_tok_s = tokenizer.tokenize(words)
    lem_s_words = [lemmatizer.lemmatize(words) for words in word_tok_s]
    lemma_s.append(" ".join(lem_s_words))
    
# Appending the lemmatized posts to the dataframe

sherlock_df["lemmatized"] = lemma_s

In [42]:
lemma_e = []

for words in elementary_df['combo']:
    word_tok_e = tokenizer.tokenize(words)
    lem_e_words = [lemmatizer.lemmatize(words) for words in word_tok_e]
    lemma_e.append(" ".join(lem_e_words))
    
# Appending the lemmatized posts to the dataframe

elementary_df["lemmatized"] = lemma_e

### Cleaning up the posts further

In [43]:
# adding words to the removal list, that are common to both subthreads and are not useful for analysis to CBS executives e.g Words like 'sherlock' and 'watson' will be common as they are the main characters.

In [44]:
#removing words that are common, changing all words to lower case to prevent counting the same word due to capital letters

In [45]:
all_stopwords = stopwords.words('english')
all_stopwords.extend(['ha','episode','episodes', 'series','spoiler','spoilers','mycroft','elementary','scene','season','seasons','show','watch', 'watching', 'watched','watched','sherlock','watson','holmes','moriarty', 'sherlocks','wa'])



In [46]:
def post_to_words(raw_post):
    # Function to convert a raw post to a string of words
    # The input is a single string (a raw post), and 
    # the output is a single string (a cleaned post)
    
   
    # 1. Convert to lower case, split into individual words.
    words = raw_post.lower().split()
    
    # 2. In Python, searching a set is much faster than searching
    # a list, so convert the stopwords to a set.
    stops = set(all_stopwords)
    
    # 3. Remove stopwords.
    meaningful_words = [w for w in words if w not in stops]
    
    # 4. Join the words back into one string separated by space, 
    # and return the result.
    return(" ".join(meaningful_words))

In [47]:
# Initialize an empty list to hold the clean posts.
clean_post_s = []


# For every post in our subreddit...
for post_s in sherlock_df['lemmatized'] :
    
    # Convert post to words, then append to clean_post.
    clean_post_s.append(post_to_words(post_s))
    
sherlock_df['lemmatized']= clean_post_s   

In [48]:
# Initialize an empty list to hold the clean posts.
clean_post_e = []


# For every post in our subreddit...
for post_e in elementary_df['lemmatized'] :
    
    # Convert post to words, then append to clean_post.
    clean_post_e.append(post_to_words(post_e))
    
elementary_df['lemmatized']= clean_post_e   

In [49]:
#dropping 'combo' column

In [50]:
sherlock_df.drop(['combo'], axis =1, inplace = True)

In [51]:
elementary_df.drop(['combo'], axis =1, inplace = True)

In [52]:
# Checking the data

sherlock_df.head()

lemmatized
0                                                                                                                                                                                      guys martin freeman
1  mark gatiss twitter response question one propswatching asib last night wondered writing picture bed mistakenly thought chinese japanese written vertical apparently written way well wrote man mark...
2  last vow s03 e03 charles augustus magnussen able read people detail pressure point first seems glass see reading without get bionic implant purpose detail shown audience like word see people read ...
3                                                                                                                                                                                              sounds like
4                                                                                                                                                                  need word describe mrs hudson ie legend

In [53]:
elementary_df.head()

lemmatized
0                                                                                                                                                                           best detective clyde love much
1  jlm exercisingthis may seem oddly specific really arbitrary trying figure jlm exercising working think crunch push ups yoga treadmill single stick boxing joan squat stay awake etc trying catalogue...
2                                                                                                                                                                stops taxi use ladder lmao s02e02 solve x
3                                                                                                                   amusing touching moment get involved character personal life know obvious purpose love
4                                           anyone got bottom reappearing green fish vase rewatch lockdown become game u find fish vase repeating background really starting bug ever solution behind joke

### Saving the cleaned data

In [54]:
sherlock_df.to_csv('../datasets/clean_sherlock.csv')

In [55]:
elementary_df.to_csv('../datasets/clean_elementary.csv')